File: backend-ruby/app/controllers/api/v1/nhat_ky_controller.rb
class Api::V1::NhatKyController < ApplicationController
  before_action :set_current_user

  def index
    return render_unauthorized unless @current_user

    records = NhatKy.includes(nguoi_dung: :vai_tro).order(ThoiGian: :desc)

    if admin?
      records = records.where(NguoiDungID: params[:user_id]) if params[:user_id].present?
    else
      records = records.where(NguoiDungID: @current_user.NguoiDungID)
    end

    render json: records.map { |record| serialize_record(record, include_user: true) }, status: :ok
  end

  def show
    return render_unauthorized unless @current_user

    record = NhatKy.includes(nguoi_dung: :vai_tro).find_by(NhatKyID: params[:id])
    return render json: { error: "Log with ID #{params[:id]} was not found." }, status: :not_found unless record

    unless admin? || record.NguoiDungID == @current_user.NguoiDungID
      return render json: { error: "You are not allowed to view this log." }, status: :forbidden
    end

    render json: serialize_record(record), status: :ok
  end

  private

  def set_current_user
    user_id = request.headers["X-User-Id"].presence
    @current_user = NguoiDung.includes(:vai_tro).find_by(NguoiDungID: user_id, TrangThai: true) if user_id
  end

  def admin?
    @current_user&.vai_tro&.TenVaiTro == "Admin"
  end

  def render_unauthorized
    render json: { error: "Khong xac dinh duoc nguoi dung dang dang nhap." }, status: :unauthorized
  end

  def serialize_record(record, include_user: false)
    user = record.nguoi_dung
    payload = {
      NhatKyID: record.NhatKyID,
      NguoiDungID: record.NguoiDungID,
      HoTen: user&.HoTen,
      VaiTro: user&.vai_tro&.TenVaiTro,
      HanhDong: record.HanhDong,
      LoaiDoiTuong: record.LoaiDoiTuong,
      DoiTuongID: record.DoiTuongID,
      MoTa: record.MoTa,
      ThoiGian: record.ThoiGian
    }

    if include_user
      payload[:NguoiDung] = user && {
        NguoiDungID: user.NguoiDungID,
        HoTen: user.HoTen,
        Email: user.Email,
        TenVaiTro: user.vai_tro&.TenVaiTro
      }
    end

    payload
  end
end

File: backend-csharp/Controllers/NhatKyController.cs
using backend_csharp.Data;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;

namespace backend_csharp.Controllers;

[Route("api/v1/nhat_ky")]
public class NhatKyController(ExamDbContext dbContext) : ApiControllerBase
{
    [HttpGet]
    public async Task<IActionResult> Index([FromQuery(Name = "user_id")] int? userId)
    {
        var currentUser = await GetCurrentUser();
        if (currentUser is null)
        {
            return Unauthorized(Error("Không xác định được người dùng đang đăng nhập."));
        }

        bool isAdmin = currentUser.VaiTro?.TenVaiTro == "Admin";
        var query = dbContext.NhatKy
            .Include(record => record.NguoiDung)
                .ThenInclude(user => user!.VaiTro)
            .AsQueryable();

        if (isAdmin)
        {
            if (userId is > 0)
            {
                query = query.Where(record => record.NguoiDungID == userId);
            }
        }
        else
        {
            query = query.Where(record => record.NguoiDungID == currentUser.NguoiDungID);
        }

        var records = await query
            .OrderByDescending(record => record.ThoiGian)
            .Select(record => new
            {
                record.NhatKyID,
                record.NguoiDungID,
                HoTen = record.NguoiDung != null ? record.NguoiDung.HoTen : null,
                VaiTro = record.NguoiDung != null && record.NguoiDung.VaiTro != null ? record.NguoiDung.VaiTro.TenVaiTro : null,
                NguoiDung = record.NguoiDung == null ? null : new
                {
                    record.NguoiDung.NguoiDungID,
                    record.NguoiDung.HoTen,
                    record.NguoiDung.Email,
                    TenVaiTro = record.NguoiDung.VaiTro != null ? record.NguoiDung.VaiTro.TenVaiTro : null,
                },
                record.HanhDong,
                record.LoaiDoiTuong,
                record.DoiTuongID,
                record.MoTa,
                record.ThoiGian,
            })
            .ToListAsync();

        return Ok(records);
    }

    [HttpGet("{id:int}")]
    public async Task<IActionResult> Show(int id)
    {
        var currentUser = await GetCurrentUser();
        if (currentUser is null)
        {
            return Unauthorized(Error("Không xác định được người dùng đang đăng nhập."));
        }

        bool isAdmin = currentUser.VaiTro?.TenVaiTro == "Admin";
        var query = dbContext.NhatKy
            .Include(item => item.NguoiDung)
                .ThenInclude(user => user!.VaiTro)
            .Where(item => item.NhatKyID == id);

        if (!isAdmin)
        {
            query = query.Where(item => item.NguoiDungID == currentUser.NguoiDungID);
        }

        var record = await query
            .Select(item => new
            {
                item.NhatKyID,
                item.NguoiDungID,
                HoTen = item.NguoiDung != null ? item.NguoiDung.HoTen : null,
                VaiTro = item.NguoiDung != null && item.NguoiDung.VaiTro != null ? item.NguoiDung.VaiTro.TenVaiTro : null,
                item.HanhDong,
                item.LoaiDoiTuong,
                item.DoiTuongID,
                item.MoTa,
                item.ThoiGian,
            })
            .FirstOrDefaultAsync();

        return record is null ? NotFound(Error($"Log with ID {id} was not found.")) : Ok(record);
    }

    private async Task<backend_csharp.Models.NguoiDung?> GetCurrentUser()
    {
        string? rawUserId = Request.Headers["X-User-Id"].FirstOrDefault();
        if (!int.TryParse(rawUserId, out int currentUserId) || currentUserId <= 0)
        {
            return null;
        }

        return await dbContext.NguoiDung
            .Include(user => user.VaiTro)
            .FirstOrDefaultAsync(user => user.NguoiDungID == currentUserId && user.TrangThai);
    }
}
